File: 백준/Gold/14500. 테트로미노/테트로미노.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
 *
 */

public class Main {

    static int[] row = {-1, 1, 0, 0};
    static int[] col = {0, 0, 1, -1};
    static boolean[][] visited;
    static int n;
    static int m;
    static int[][] grid;
    static int ele = 0;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        grid = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                grid[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        visited = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                visited[i][j] = true;
                dfs(i, j, 0, grid[i][j]);
                visited[i][j] = false;
            }
        }

        // ㅗ 모양
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                checkF(i,j);
            }
        }

        System.out.println(ele);


    }

    private static void checkF(int i, int j) {
        // ㅏ 모양
        if(i+2<n && j+1<m){
            int sum = grid[i][j] + grid [i+1][j]+grid[i+2][j]+grid[i+1][j+1];
            ele = Math.max(sum, ele);
        }
        // ㅓ 모양
        if(i+2<n && j-1>=0){
            int sum = grid[i][j] + grid [i+1][j]+grid[i+2][j]+grid[i+1][j-1];
            ele = Math.max(sum, ele);
        }
        // ㅜ 모양
        if(i+1<n && j+2<m){
            int sum = grid[i][j] + grid [i][j+1]+grid[i][j+2]+grid[i+1][j+1];
            ele = Math.max(sum, ele);
        }
        // ㅗ 모양
        if(i-1>=0 && j+2<m){
            int sum = grid[i][j] + grid [i][j+1]+grid[i][j+2]+grid[i-1][j+1];
            ele = Math.max(sum, ele);
        }
    }

    private static void print(boolean[][] visited) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (visited[i][j]) {
                    System.out.print("1 ");
                } else {
                    System.out.print("0 ");
                }
            }
            System.out.println();
        }
    }

    private static void dfs(int r, int c, int num, int sum) {
        if (num >= 3) {
//            System.out.println("-------------------");
//            System.out.println(sum);
//            print(visited);
            ele = Math.max(ele, sum);
            return;
        }


        for (int k = 0; k < 4; k++) {
            int nr = r + row[k];
            int nc = c + col[k];
            if (nr >= 0 && nr < n && nc >= 0 && nc < m && !visited[nr][nc]) {
                visited[nr][nc] = true;
                dfs(nr, nc, num + 1, sum + grid[nr][nc]);
                visited[nr][nc] = false;
            }
        }

    }
}

File: 백준/Gold/14500. 테트로미노/테트로미노.py
'''
문제 설명
    모든점에서 가능한 depth가 4의 합들을 구해서 최댓값 갱신
    모양은 depth 4의 모든 경우의 수임
    그런데 ㅗ 모양이 돌아오는 로직이 필요
    -> 하드코딩 할거냐 vs 돌아올거냐
    -> 하드코딩..

필요한 메서드
    dfs

4 6
100 100 100 1 1 1
1 100 1 1 1 1
1 1 1 1 1 1
1 1 1 1 1 1


백준은 틀림 뭐지....
'''

n,m = map(int, input().split())

grid = [list(map(int, input().split())) for i in range(n)]

ans = 0

visited = [[False]*m for i in range(n)]
row = [-1,1,0,0]
col = [0,0,1,-1]
def dfs(r,c,depth,sm):
    global ans
    if depth == 4:
        ans = max(ans, sm)
        # if sm == 72:
        #     print(r,c) # 디버깅하려고 넣어놓은게 있어서 틀림 ㅠㅠ
        return

    for k in range(4):
        nr = r+row[k]
        nc = c+col[k]

        if not(0<=nr<n and 0<=nc<m) or visited[nr][nc]:
            continue

        visited[nr][nc] = True
        dfs(nr,nc,depth+1,sm+grid[nr][nc])
        visited[nr][nc] = False



for i in range(n):
    for j in range(m):
        visited[i][j] = True
        dfs(i,j,1,grid[i][j]) # 위치와 depth, 합
        visited[i][j] = False

# ㅏ 모양
for i in range(n-2):
    for j in range(m-1):
        ele_sum = grid[i][j]+grid[i+1][j]+grid[i+2][j]+grid[i+1][j+1]
        ans = max(ans, ele_sum)

# ㅓ 모양
for i in range(n-2):
    for j in range(1,m):
        ele_sum = grid[i][j]+grid[i+1][j]+grid[i+2][j]+grid[i+1][j-1]
        ans = max(ans, ele_sum)

# ㅜ 모양
for i in range(n-1):
    for j in range(1,m-1):
        ele_sum = grid[i][j] + grid[i][j-1] + grid[i][j+1] + grid[i + 1][j]
        ans = max(ans, ele_sum)

# ㅗ 모양
for i in range(n-1):
    for j in range(1,m-1):
        ele_sum = grid[i][j] + grid[i+1][j-1] + grid[i+1][j] + grid[i + 1][j + 1]
        ans = max(ans, ele_sum)

print(ans)